### DISCUSSION TITLE
Failed to delete record

### DISCUSSION THREAD

**Discussion by synapse** [pocketbase/pocketbase#1220] [Q&A]
I am getting the following error message when I'm trying to delete the record. 

`Failed to delete record. Make sure that the record is not part of a required relation reference.`

I understand the reason why is showing the message but it is not clear which of the fields are blocking the deletion. All the relational fields in this table have the option `Delete record on relation delete` set to `False`.
There are not other tables that have references to this one.

**Comment by ganigeorgiev** [comment]
What version of PocketBase are you using? [PocketBase v0.9.1](https://github.com/pocketbase/pocketbase/releases/tag/v0.9.1) comes with several fixes, some of which related to the record deletion.

You could also start the executable with `--debug` and eventually see if there is a more detailed log in the console.

**Reply by synapse** [reply]
I'm using 0.0.9, I'll get back a you with my findings after I try out 0.9.1

**Reply by ganigeorgiev** [reply]
If after upgrade you are still getting the same error, then it is possible that the "Nonempty/Required" constraint of the relation field is blocking the deletion, since when "Cascade delete" is set to `false` we try to "unset" the relation value if possible (aka. empty string or array depending on the "Max Select" option).

**Reply by synapse** [reply]
Yes, so after the upgrade the problem was the same. From the debug messages I saw this one:

`2022/12/09 17:05:53 SQL logic error: ambiguous column name: parent (1)`

renaming that column to something else fixed the issue. Are there other reserved column names that might cause problems?

**Reply by ganigeorgiev** [reply]
No, this is probably a bug introduced with v0.9.0. I'll test it now.

**Reply by ganigeorgiev** [reply]
@synapse Could you provide more details on your collections structure? Or the sql log before the "SQL logic error"?

**Reply by synapse** [reply]
Sure, this is the log just after I hit delete on the record:

```
[3.00ms] SELECT {{_admins}}.* FROM `_admins` WHERE `id`='zgzolnven3m3msn' LIMIT 1
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = '6qcj7li7hfjvaio' OR LOWER([[name]])='6qcj7li7hfjvaio' LIMIT 1
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = '6qcj7li7hfjvaio' OR LOWER([[name]])='6qcj7li7hfjvaio' LIMIT 1
[0.00ms] SELECT `tasks`.* FROM `tasks` WHERE `tasks`.`id`='5w1axmi0q1krjnl' LIMIT 1
[0.00ms] SELECT {{_collections}}.* FROM `_collections`
[0.00ms] DELETE FROM `tasks` WHERE `id`='5w1axmi0q1krjnl'
[0.00ms] SELECT DISTINCT `timelogs`.* FROM `timelogs` LEFT JOIN json_each(CASE WHEN json_valid([[task]]) THEN [[task]] ELSE json_array([[task]]) END) as {{__je__}} WHERE (NOT (`timelogs`.`id`='5w1axmi0q1krjnl')) AND (`__je__`.`value`='5w1axmi0q1krjnl')
[0.00ms] SELECT DISTINCT `tasks`.* FROM `tasks` LEFT JOIN json_each(CASE WHEN json_valid([[subtasks]]) THEN [[subtasks]] ELSE json_array([[subtasks]]) END) as {{__je__}} WHERE (NOT (`tasks`.`id`='5w1axmi0q1krjnl')) AND (`__je__`.`value`='5w1axmi0q1krjnl')
[0.00ms] SELECT DISTINCT `tasks`.* FROM `tasks` LEFT JOIN json_each(CASE WHEN json_valid([[parent]]) THEN [[parent]] ELSE json_array([[parent]]) END) as {{__je__}} WHERE (NOT (`tasks`.`id`='5w1axmi0q1krjnl')) AND (`__je__`.`value`='5w1axmi0q1krjnl')
2022/12/09 17:13:06 SQL logic error: ambiguous column name: parent (1)
[0.00ms] SELECT {{_admins}}.* FROM `_admins` WHERE `id`='zgzolnven3m3msn' LIMIT 1
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = '6qcj7li7hfjvaio' OR LOWER([[name]])='6qcj7li7hfjvaio' LIMIT 1
[0.00ms] SELECT COUNT(*) FROM (SELECT `tasks`.`id` FROM `tasks`)
[0.00ms] SELECT `tasks`.* FROM `tasks` ORDER BY [[tasks.created]] DESC LIMIT 30
```
If you need any other info, let me know.

**Reply by ganigeorgiev** [reply]
This is a bug and it was introduced with v0.9.0 because `json_each` has a `parent` field and it conflicts with the rel column. We'll need to prefix it. I'll try to add a test and will make a new release shorty after that.

**Reply by ganigeorgiev** [reply]
@synapse A fix was released in [v0.9.2](https://github.com/pocketbase/pocketbase/releases/tag/v0.9.2). Please let me know if you are still experiencing issues after the update.

**Reply by synapse** [reply]
hey @ganigeorgiev all good now, works as expected 👍 .
Thanks a lot for the quick fix!

**Reply by Jeffreyon** [reply]
Hi i'm currently experiencing this issue with v12.2

Is there a fix for it?

**Reply by ganigeorgiev** [reply]
@Jeffreyon 
What exactly is the returned error?
Could you elaborate a little more on your collections structure?
If the error is the same as the above, could you confirm that your records don't have a required relation reference anywhere?

**Comment by Jeffreyon** [comment]
I don't really understand what "required relation reference" means but
here's how my collections are related

There's a 'users' collection, users can ask a question which gets stored in
the 'questions' collection and they can get replies on their questions
which gets stored in a 'replies' collection

A reply has a one to one relation to a question, which has a one-to-one
relation to a user.


I wanted to delete a user, but I couldn't

Although I just tried reversing the operation, I first deleted the replies,
then the question and finally the user and It all worked

I don't know what I'm doing wrong with my collection structure. I also set
Cascade to false on all the relation fields in 'questions' and 'replies'
collection

**Reply by ganigeorgiev** [reply]
A "required relation reference" means that a `relation` field cannot be empty.

If `replies` or `questions` have a "Nonempty" (aka. required) `relation` field then you can't delete a user because on user delete we clear the related field values to their zero-defaults (empty string or array). The "Cascade" option is only for deleting the entire record, aka. on user delete -> delete also the related record, which I assume should be the expected behavior in your case and you should set it to `true` (this is why it also works when you delete first the replies and then the question, but if you set the Cascade to true, then this will happen automatically).

If that's not the case, please export your collections structure as json and attach it here (Admin UI > Settings > Export collections).

**Reply by synapse** [reply]
I can relate to this problem myself where I know there are relations between tables but it would be nice to have like a more verbose error message saying something like: `record cannot be deleted because of this relation on field and this table`. I am using multi relational table where I might have more then 5 relations and it's gets tricky to catch which one is blocking the deletion. In fact I've adopted a method to just add a new column called `deleted` and an API filter that goes with it to mark something as deleted. This helped me when I was trying to delete a record with circular references and deletion order.

**Reply by ganigeorgiev** [reply]
@synapse There is already a more detailed error message in the Admin UI > Logs and/or when you start PocketBase with `--debug`. Eventually, we may propagate and print it also to the end user but someone could argue that this may lead to involuntary information disclosure about the internals of your data.

**Reply by synapse** [reply]
Yes that's true. I was only talking about a more detailed message for the admin panel and not for the API response.
